android/samples/sample-gltf-viewer/src/main/java/com/google/android/filament/gltf/MainActivity.kt

```diff
@@ -74,6 +111,126 @@ class MainActivity : Activity() {
         modelViewer.view.bloomOptions = bloomOptions
     }
 
+    private fun bindPreview(cameraProvider: ProcessCameraProvider) {
+        val preview : Preview = Preview.Builder().build().also {
+            it.setSurfaceProvider(previewView.createSurfaceProvider())
+        }
+
+        val cameraSelector : CameraSelector = CameraSelector.Builder()
+                .requireLensFacing(CameraSelector.LENS_FACING_BACK)
+                .build()
+
+        val imageAnalysis = ImageAnalysis.Builder()
+                .setTargetResolution(Size(1280, 720))
+                .setBackpressureStrategy(ImageAnalysis.STRATEGY_KEEP_ONLY_LATEST)
+                .build()
+
+        imageAnalysis.setAnalyzer(AsyncTask.THREAD_POOL_EXECUTOR, {
+            imageProxy -> scanQRCode(imageProxy)
+        })
+
+        try {
+            cameraProvider.unbindAll()
+            cameraProvider.bindToLifecycle(this as LifecycleOwner, cameraSelector, imageAnalysis, preview)
+        } catch (exc: Exception) {
+            Log.e(TAG, "Preview binding failed", exc)
+        }
+    }
+
+    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
+        menuInflater.inflate(R.menu.main_menu, menu)
+        return true
+    }
+
+    override fun onOptionsItemSelected(item: MenuItem) = when (item.itemId) {
+        R.id.action_qrcode -> {
+            if (previewView.visibility == View.GONE) {
+                surfaceView.visibility = View.GONE
+                previewView.visibility = View.VISIBLE
+            } else {
+                surfaceView.visibility = View.VISIBLE
+                previewView.visibility = View.GONE
+            }
+            true
+        }
+        else -> {
+            super.onOptionsItemSelected(item)
+        }
+    }
+
+    @SuppressLint("UnsafeExperimentalUsageError")
+    private fun scanQRCode(imageProxy: ImageProxy) {
+        val mediaImage = imageProxy.image ?: return
+        if (previewView.visibility != View.VISIBLE) {
+            imageProxy.close()
+            return
+        }
+        val options = BarcodeScannerOptions.Builder().setBarcodeFormats(Barcode.FORMAT_QR_CODE).build()
+        val image = InputImage.fromMediaImage(mediaImage, imageProxy.imageInfo.rotationDegrees)
+        val scanner = BarcodeScanning.getClient(options)
+        scanner.process(image)
+                .addOnSuccessListener { barcodes ->
+                    for (barcode in barcodes) {
```

Comment:
If multiple barcodes are detected, does it make sense to call `downloadFromBarcode` for each? Perhaps we just want to download the first one we find.

Reply:
Good catch, I will simplify this.